### ISSUE TITLE
Need a big optimization for Windows

### ISSUE THREAD

**Issue by OshotOkill** [yarnpkg/yarn#990]
**Do you want to request a _feature_ or report a _bug_?**

Feature

**What is the current behavior?**

I've tested a lot about the installation speed between MacOS and Windows. According to the results it seems that yarn has far less optimizations for Windows. e.g. Here are the comparisons of installing `react-native` :

<br />

> Test Machines: 
> - ThinkPad X1 Carbon 4, 1TB PCI-E SSD, 16GB Memory
> - MacBook Air 2014, 256GB SSD, 4GB Memory
> 
> No cache & same network environment

<br />

**MacOS**

npm@3.10.9:  **1m 31s**

![2016-10-13 17 52 24](https://cloud.githubusercontent.com/assets/6079112/19346419/ca8abd58-9175-11e6-80af-0ec1a3baef7e.png)

yarn@0.15.1:  **39s**

![2016-10-13 17 54 53](https://cloud.githubusercontent.com/assets/6079112/19346420/ca8b7c98-9175-11e6-9776-b864db49d10f.png)

<br />

**Windows**

npm@3.10.9: **2m 24s**

![2](https://cloud.githubusercontent.com/assets/6079112/19346739/8b14e1ec-9177-11e6-9437-642b6d4e60c1.PNG)

yarn@0.15.1: **2m 19s**

![1](https://cloud.githubusercontent.com/assets/6079112/19346742/8ed1370e-9177-11e6-9135-56228dac77c9.png)

<br />

So, it seems yarn has no advantage over npm on Windows. Did anyone face with this appearance?

**Please mention your node.js, yarn and operating system version.**
nodejs:  6.8.0
yarn:      0.15.1
OS:        Windows 10 14393.321 & MacOS 10.12

**Comment by atkawa7** [comment]
+1

**Comment by Daniel15** [comment]
Hi @OshotOkill! Thanks for trying Yarn. Are you using Cygwin or WSL ("Bash on Ubuntu on Windows")? Both are known to have pretty bad disk IO performance.

Also, React Native has a *huge** number of files so copying them into `node_modules` is pretty slow, and disk IO for lots of little files on Windows is generally slower than Mac OS (which itself is slower than Linux ext4). We do have a task to experiment with hardlinks (#499) which should improve perf in this scenario. 

> No cache & same network environment

The main improvement with Yarn is when you have a warm cache (ie. after you've installed a package at least once).

**Comment by OshotOkill** [comment]
@Daniel15 Nope, I'm not using Cygwin/MinGW/MSYS2 or WSL (the latter fails due to a knotty bug).

According to your description I can assume the problem is caused by the file system (NTFS) right ? Even if a warm cache exists, the copying process still runs much slower than MacOS.

Hope the dev teams can come up with a solution ASAP. Thanks.

**Comment by fooey** [comment]
I'm seeing the same.

Do install, wipe node_modules, install

MacBookPro takes 17 seconds, my Windows machine takes 122 seconds.

**Comment by cpojer** [comment]
Somebody pointed out this might be related to anti-virus software scanning node_modules and the global yarn cache continuously. Can you try disabling it for those folders?

**Comment by OshotOkill** [comment]
@cpojer I guess they are right . I don't have any anti-virus software on my machine expect the pre-installed Windows Defender, so I banned the scanning of global cache folder & my project folder and did some test:

**Default:  128.08s**

![2](https://cloud.githubusercontent.com/assets/6079112/19390752/0a2c71f4-925c-11e6-8e52-9189dc8cd481.PNG)

<br />

**No scanning of cache folder: 104.43s**

![3](https://cloud.githubusercontent.com/assets/6079112/19390776/2354e1a2-925c-11e6-8183-9b230a427d84.PNG)

<br />

**No scanning of project folder:  78.28s**

![5](https://cloud.githubusercontent.com/assets/6079112/19390788/308cd136-925c-11e6-9be8-10900b657aae.PNG)

<br />

**No scanning of cache folder & project folder:  53.48s**

![4](https://cloud.githubusercontent.com/assets/6079112/19390797/3e1d17ac-925c-11e6-860b-93660d02f53f.PNG)

<br />

Though it's slower than Mac for 10+s, it has a significant boost.

This should be informed from the official docs I think.

**Comment by Daniel15** [comment]
> Somebody pointed out this might be related to anti-virus software scanning node_modules and the global yarn cache continuously. 

Good catch! I totally forgot about this as I already have `c:\src` whitelisted on my computers.

@OshotOkill - Would you like to submit a pull request adding a note about antivirus apps to the website? Here's the file you'd need to edit: https://github.com/yarnpkg/website/blob/master/en/docs/_installations/windows.md (you can edit it directly on Github). It'd be appreciated 😄

**Comment by fooey** [comment]
I wasn't as meticulous as @OshotOkill, but I added exceptions for my source and my node install folder, and then specifically exempted the yarn, npm and node binaries and now my fresh install time on Windows is down to 50 seconds from 122 seconds.

**Comment by OshotOkill** [comment]
@Daniel15 PR is ready. Apologize for my poor English.

**Comment by OshotOkill** [comment]
PR has been merged. Close this issue.

**Comment by voliva** [comment]
This is still painfully slow on windows, even deactivating anti-virus and Windows defender. I don't think it's just an environment issue (like this anti-virus solution) but it looks like yarn tries to copy all the files, 1-by-1 even if you install some unrelated dependency.

Why not just delete/copy the files that need to change?

**Comment by amcsi** [comment]
I have created a StackOverflow article about this too: http://stackoverflow.com/questions/40566222/yarn-5x-slower-on-windows

By the way in my (dual-booted) Ubuntu benchmarks I was using the same NTFS drive as the one Windows normally runs on; and it's still fast there.

**Comment by bestander** [comment]
Adding `node.exe` to Windows Defender exclusions gave me a huge performance boost http://126kr.com/article/1884rsed7l

**Comment by amcsi** [comment]
I'll definitely try this out!

**Comment by amcsi** [comment]
It did seem to improve the speed a bit 212 -> 170 seconds
So it seems to help, but it could still be improved, because it's still more than 3x slower than in Linux

**Comment by bestander** [comment]
Another issue I have noticed - Indexing service on Windows tries to index every file in node_modules.
I don't really need it at all, so I disabled it http://www.softwareok.com/?seite=faq-Windows-10&faq=53 and it gave another performance boost.

**Comment by amcsi** [comment]
My windows isn't set to index the path in question, so that still doesn't solve the issue.

**Comment by Altiano** [comment]
So to sum up there are 4 ways to improve performance:

-  Whitelist project folder from AV
-  Whiteilst the Yarn cache directory ((%LocalAppData%\Yarn)) from AV
-  Adding `node.exe` to Windows Defender exclusions
-  Disabling Indexing service on Windows on `node_modules` folder

**Comment by amcsi** [comment]
@Altiano yes, but it's still not enough to get performance even close to Mac/Linux

**Comment by michaellashinsky** [comment]
Seems kinda sketchy that you'd have to disabled AV or indexing on directories to make yarn as fast or faster than npm.  After all, you don't have to do this for npm.  I decided to give yarn a shot because it stated it was fast and the offline installs made coding with out a network connection a plausible thing. Is there no way to optimize the linking?

**Comment by OshotOkill** [comment]
According some issues who relate here and the comments above, I'd like to reopen this issue in order to gather some other solutions.

Personally I suggest to list the hardware configurations about your test machine and upload some related pics. There could be many other irrelevant elements that make a big difference between platforms rather than Yarn itself, i.e. the benchmark performance of the SSD on a MacBook is usually much better than a Windows machine.

**Comment by amcsi** [comment]
@OshotOkill like I said earlier, I got 3.5x slower performance on Windows vs Linux having indexes and windows defender disabled for the relevant directories, on the same regular PC on the same _ntfs_ drive. That that even on ntfs it's much faster on Linux says a lot I think.

**Comment by bestander** [comment]
Let's get to the reason for this.
Could it be NTFS working slower on a large number of files being moved during install?

Can anyone share a way to repro this on a single machine?
For example, a particular package.json installed on a Windows laptop takes X seconds but running in VirtualBox Ubuntu X-20% seconds.

**Comment by OshotOkill** [comment]
@amcsi @bestander  As is often the case, EXT4/XFS are faster while copying large amounts of small files. However, NTFS is not that much slower. I just cleaned the cache and tested again by using the latest version of Yarn and Node (0.19.1 & 7.5.0):

 
![image](https://cloud.githubusercontent.com/assets/6079112/22623344/69e84a3e-eb92-11e6-91b4-1188c68fe1aa.png)

The result is really close to a MacBook while installing `react-native`. All I did was just whitelist the related folders and Node.exe process.

**Comment by nozzlegear** [comment]
I was having this problem myself until I whitelisted the node.exe and yarn.exe processes in Windows Defender, along with my project directory. I haven't disabled search indexing at all, nor have I whitelisted the Yarn cache directory. Install times went from 190+ seconds on a medium-sized project to about 25 seconds from a clean cache. My Ubuntu machine is just a little bit faster than that, but only by 5-10 seconds.

![Fresh Yarn install](https://i.imgur.com/BrI7so5.png)

**Comment by keawade** [comment]
I just did some quick tests on my own system. I've got Linux Mint and Windows 10 dual booted off the same SSD. I cleaned my yarn cache, deleted `node_modules` and ran `yarn` on this [vue project](https://github.com/keawade/keawade.github.io/tree/vue).

## Linux Mint: _12.22s_

![yarnlinuxmint](https://cloud.githubusercontent.com/assets/7308850/22623759/40a037de-eb2b-11e6-84ef-0224e545e645.png)

## Windows 10 (No white listing): _64.32s_

![yarnwindows10](https://cloud.githubusercontent.com/assets/7308850/22623761/5c1b2028-eb2b-11e6-8d3e-9ccfb400d54d.PNG)

## Windows 10 (With white listing): _42.58s_

![yarnwindows10_withexclusions](https://cloud.githubusercontent.com/assets/7308850/22623762/65298510-eb2b-11e6-9a26-f4b83149c50c.PNG)

These were the Windows Defender exclusions I had active:
![yarnwindows10_exclusions](https://cloud.githubusercontent.com/assets/7308850/22623763/69ac591e-eb2b-11e6-9896-9ac7cb3fb4d2.PNG)

While white listing did seem to have a significant effect, it still didn't come close to matching the speed on Linux.

**Comment by nozzlegear** [comment]
@keawade I had 26.48s to install your project from a clean cache, and 13.58s to install it with the cache.

![keawade.github.io](https://i.imgur.com/GLGUGIJ.png)

Just spitballing here, I'm using the Yarn.cmd from the MSI installer and it looks like you're using Yarn installed from NPM. I wonder if there's maybe a discrepancy between them?

**Comment by keawade** [comment]
@nozzlegear While that might be possible, I think that is less likely than it being due to differing internet connections.

**Comment by bestander** [comment]
We need to eliminate network from this.
Currently I can test [this](https://github.com/keawade/keawade.github.io/tree/vue) repo on a latest Windows 10 with Linux on Windows enabled.
Both via CMD and Bash with prime caches installation takes about 27-29 seconds.

@keawade, can you run the same test with node_modules removed but caches in place?

I can't install a second OS on the device I have yet.
Can anyone check if running Windows and Linux in a Virtual box give different results?

I've built current master with timestamps https://github.com/yarnpkg/yarn/releases/download/v0.21.0-pre/yarn-0.21.0-0.js

Can you use it for installing with `--verbose` flag?

E.g.
```
node /Users/bestander/work/yarn/artifacts/yarn-0.21.0-0.js install --verbose
```

**Comment by keawade** [comment]
## Data without cleaning caches

_Note: This data is being recorded on a dual booted system. All hardware is identical for these tests._

| OS         | Avg Time | Normalized |
|------------|----------|------------|
| Linux Mint | 5.598s   | 1.000      |
| Windows 10 | Pending  | Pending    |

### Raw Linux Mint Data

`[5.47, 5.40, 5.84, 5.96, 5.55, 5.48, 5.40, 5.57, 5.81, 5.50]`

### Raw Windows 10 Data

I will update this post in a few minutes with the Windows 10 data and update the table.

**Comment by keawade** [comment]
@bestander, I've updated the previous post with the Windows data.

**Comment by bestander** [comment]
Great, thanks for more data.
Can you try the --verbose version with yarn.js with time stamps for both OS?
It would give us a good idea where time is spent.

**Comment by keawade** [comment]
Whew, that is a lot of logging! Do you want 10 runs for each OS / white list combination or is one for each good enough?

**Comment by keawade** [comment]
@bestander Here you go! One of each.

Side note: Turns out if you try to upload ~30mb of raw text to a single gist collection you get an nginx 405 error. 😆

[LinuxMint.txt](https://gist.github.com/keawade/65133faeeff393eb5f29e0fe0f3c603f)
[Windows 10 with **exclusions** and with **clean**](https://gist.github.com/keawade/400fe0d169e17ae76e97377f95d4d85d)
[Windows 10 with **exclusions** and without _clean_](https://gist.github.com/keawade/86fbfd8f40412c5b23e9ea20344e1f80)
[Windows 10 with **clean** and without _exclusions_](https://gist.github.com/keawade/a0aead83a85500d431bb57674ff29775)
[Windows 10 without _exclusions_ and _without_ clean](https://gist.github.com/keawade/a4d13d352d378ae2dd3c1c137a74d0ff)

**Comment by Daniel15** [comment]
> Turns out if you try to upload ~30mb of raw text to a single gist collection you get an nginx 405 error. 😆

You could compress the files (bzip2 or 7-Zip) and attach them here... Plain text compresses very well :)

**Comment by keawade** [comment]
@Daniel15 Good point, here are the compressed files: [VerboseLogs.tar.gz](https://github.com/yarnpkg/yarn/files/753312/VerboseLogs.tar.gz)

**Comment by bestander** [comment]
1 run would be fine :)

**Comment by bestander** [comment]
I compared LinuxMint.txt vs Windows10NoClean.txt

Linux:
- linking phase starts at 1.156 seconds
- all folders inside node_modules created at 1.968
- last file copied at 3.873 seconds
- builds are done in another 3 seconds

Windows
- linking phase starts at 2.779 seconds
- all folders inside node_modules created at 4.83
- last file copied at 32.853
- builds are done in another 3 seconds

Obviously verbose logging affects execution time on Windows (12 -> 35 seconds) but not on Linux (same 6 seconds).

**Comment by bestander** [comment]
From the benchmarks I found on the internet Linux EXT3 FS usually outperforms NTFS when a lot of files are copied.
I wonder if this is the limit we have to face.

@keawade, are the speeds different when using npm@3 on Windows and Linux?

A few ideas:
- Windows may be bad at concurrent copy, we copy files in 4 threads. Maybe do it single threaded?
- Use robocopy wrapper in Windows https://github.com/mikeobrien/node-robocopy
- we use readstream.pipe.writestream to copy files, maybe it is inefficient on Windows

If you are eager to experiment, replace `4` with `1` in https://github.com/yarnpkg/yarn/blob/master/src/util/fs.js#L322 and see if single threaded copying gets faster on windows

**Comment by keawade** [comment]
## Thread Tests

Per @bestander's request, I forked `yarnpkg/yarn` and modified line 322 of `src/util/fs.js`, replacing the `4` with a `1`. I then used `yarn run build` to build the project and [ran 10 tests](https://gist.github.com/keawade/493194fca7e79a22f0f9e82c9d5bdc62) with that build using the `yarn.cmd` that was compiled by the build. These are the results.

|                            | Avg Time | Normalized |
|----------------------------|----------|------------|
| Windows 10 (w/ White list) | 12.119s  | 1.00000    |
| Single copy thread         | 16.927s  | 1.39673    |
| Single copy thread + Clean | 42.268s  | 3.48775    |

_Avg Time is the average across a set of 10 tests_

It looks like using only a single thread to copy the files results in slightly slower install times.

### Raw Data

#### Windows 10 (w/ White list)

This data is from a [previous test](https://github.com/yarnpkg/yarn/issues/990#issuecomment-277542184)

#### Single Copy Thread

[15.72, 17.43, 15.16, 17.21, 17.83, 17.47, 16.68, 16.58, 16.93, 18.26]

#### Single Copy Thread + Clean

[37.68, 40.10, 43.20, 46.18, 40.84, 40.58, 39.69, 47.93, 42.45, 44.03]

**Comment by bestander** [comment]
Thanks, @keawade.
Can you verify my assumption that NTFS might be slower at copying large number of files than Linux FS?

Measure copying via terminal full installed node_modules to another location in both Linux Mint and Windows 10, please.

**Comment by pongo** [comment]
It is also necessary to test the copy using robocopy with the option `/mt` (multi-threaded copies)

**Comment by kumarharsh** [comment]
I'd also like to report a possibly reported bug, wherein every single `yarn add` or `yarn remove` takes about 30-40 minutes. It apparently copies ALL the dependencies again, and since I'm on Windows, this takes a long time. See linked issue:

https://github.com/yarnpkg/yarn/issues/2460

**Comment by OshotOkill** [comment]
@kumarharsh #2458 It  took me 28s to finish the installation.

![image](https://cloud.githubusercontent.com/assets/6079112/22649503/2b11c2e8-ecb6-11e6-96d0-bd5964404dff.png)

Also I must mention that do not forget to whitelist the project folders as well, not only the cache.

**Comment by keawade** [comment]
## Copy Tests

I used [this script](https://gist.github.com/keawade/493194fca7e79a22f0f9e82c9d5bdc62#file-copy-test-ps1) to run 10 iterations of copies on both Linux Mint and Windows 10. I copied [this repo](https://github.com/keawade/keawade.github.io/tree/vue) after running `yarn` in the directory. These are my results.

| OS         | Avg Time   | Normalized |
|------------|------------|------------|
| Linux Mint |  1527.4620 |  1.00000   |
| Windows 10 | 53676.3155 | 35.14085   |

That time difference is crazy. These copies were done copying files from one location to another on _the same SSD_.

### Raw Data

#### Linux Mint

```
TotalMilliseconds
-----------------
        1515.3961
        1513.9469
        1540.3275
        1527.2777
        1514.6029
        1521.3711
        1512.0628
        1547.8331
        1518.1499
        1563.6521
```

#### Windows 10

```
TotalMilliseconds
-----------------
       55729.4968
       55915.5972
       53427.5155
       51624.6760
       52191.4177
       53556.4542
       53562.5533
       53527.9015
       53610.6127
       53616.9302
```

I don't have time right now to test robocopy but I can get that data this evening after work.

**Comment by keawade** [comment]
## Robocopy Test

I used [this script](https://gist.github.com/keawade/493194fca7e79a22f0f9e82c9d5bdc62#file-copy-test-ps1) to run 10 iterations of copies on both Linux Mint and Windows 10. I copied [this repo](https://github.com/keawade/keawade.github.io/tree/vue) after running `yarn` in the directory. These are my results.

| OS                    | Avg Time   | Normalized |
|-----------------------|------------|------------|
| Linux Mint            |  1527.4620 |  1.00000   |
| Windows 10            | 53676.3155 | 35.14085   |
| Windows 10 (Robocopy) | 58089.7457 | 38.03024   |

Robocopy performed slightly worse than a regular copy.

### Raw Data

The Linux Mint and Windows 10 values are from the [previous tests](https://github.com/yarnpkg/yarn/issues/990#issuecomment-277689686)

```
TotalMilliseconds
-----------------
       56935.3304
       58234.8084
       57838.7956
       56731.7850
       58380.1805
       58097.6040
       59161.0365
       59062.9404
       58363.5527
       58091.4234
```

**Comment by bestander** [comment]
@keawade, can you verify that file indexing and Defender don't interfere with the copy?
Afaik it can get involved even for a cp command.

Check what is active in Task Manager when the copy is done.
And maybe just turn off those services for a test

**Comment by keawade** [comment]
## Indexing and Defender Tests

I performed tests under the following conditions:

* With disabled Windows Defender
* With disabled Windows Indexing service
* With _both_ disabled Windows Defender and Windows Indexing service
* With _both_ disabled Windows Defender and Windows Indexing service _and_ cleaning the Yarn cache

To disable Windows Defender, I toggled off `Real-time protection` under the [Windows Defender settings panel](https://cloud.githubusercontent.com/assets/7308850/22678112/3dcc446a-ecbd-11e6-8cc9-1a5164264858.PNG).

To disable Windows indexing, I stopped the Windows Search service in the [Services control panel](https://cloud.githubusercontent.com/assets/7308850/22678130/64a4bafe-ecbd-11e6-9fe6-c48c3fe5a520.PNG).

_Note: When Windows Defender was enabled, no exclusions were listed_

I used [this script](https://gist.github.com/keawade/493194fca7e79a22f0f9e82c9d5bdc62#file-copy-test-ps1) to run 10 iterations of copies on both Linux Mint and Windows 10. I copied [this repo](https://github.com/keawade/keawade.github.io/tree/vue) after running `yarn` in the directory. These are my results.

### Summary

It looks like while Windows indexing (Search service) does have an impact on copy operations and Yarn, the larger impact comes from Windows Defender.

### Copying

| OS                                         | Avg Time   | Normalized |
|--------------------------------------------|------------|------------|
| Linux Mint                                 |  1527.4620 |  1.00000   |
| Windows 10 (No defender)                   |  7301.4307 |  4.78011   |
| Windows 10 (No indexing)                   | 10307.0794 |  6.74787   |
| Windows 10 (No defender, no indexing)      |  7044.1393 |  4.61166   |
| Windows 10 Robo (No defender, no indexing) | 10094.8358 |  6.60889   |

Indexing Fully disabling indexing and antivirus provides a huge boost to performance when copying files.

### Yarn

Since the results above were so pronounced, I figured we could probably use data on Yarn's performance under these conditions as well.

I used [this script](https://gist.github.com/keawade/493194fca7e79a22f0f9e82c9d5bdc62#file-yarn-test-ps1) to run 10 iterations of `yarn` on both Linux Mint and Windows 10. I cloned [this repo](https://github.com/keawade/keawade.github.io/tree/vue) and ran `yarn` in the directory.

| OS                                          | Avg Time | Normalized |
|---------------------------------------------|----------|------------|
| Linux Mint                                  |   5.5980 | 1.00000    |
| Windows 10 (No defender)                    |  16.5450 | 2.95552    |
| Windows 10 (No indexing)                    |  38.5170 | 6.88049    |
| Windows 10 (No defender, no indexing)       |  16.8490 | 3.00982    |
| Windows 10 Clean (No defender, no indexing) |  30.7730 | 5.49714    |


### Raw Data

The Linux Mint values are from the [previous](https://github.com/yarnpkg/yarn/issues/990#issuecomment-277542184) [tests](https://github.com/yarnpkg/yarn/issues/990#issuecomment-277689686).

#### Windows 10 Copy-Item

```
[7053.7702, 7163.6924, 7081.5366, 7131.2731, 6887.5165, 6960.7251, 6999.6528, 7051.1932, 7046.8592, 7065.1741]
```

#### Windows 10 Robocopy

```
[10096.4991, 10290.1073, 10350.6061, 9999.0552, 10294.0660, 10024.2568, 9949.6786, 9878.1346, 9801.2121, 10264.7418]
```

#### Windows 10 Yarn

```
[16.81, 16.23, 16.29, 16.48, 19.03, 16.27, 17.64, 16.64, 16.05, 17.05]
```

#### Windows 10 Yarn Clean

```
[47.46, 27.83, 28.31, 27.87, 28.90, 30.70, 31.17, 27.97, 28.77, 28.75]
```

#### Windows 10 Yarn Indexing Disabled

```
[38.47, 38.63, 38.37, 38.82, 38.05, 38.54, 38.44, 37.90, 39.02, 38.93]
```

#### Windows 10 Copy Indexing Disabled

```
[10222.4855, 10063.3654, 10152.2953, 10151.6155, 10316.7628, 10705.8277, 10199.5391, 10624.1961, 10308.2336, 10326.4731]
```

#### Windows 10 Yarn Windows Defender Disabled

```
[17.03, 16.21, 16.76, 16.43, 16.19, 16.71, 16.23, 16.30, 17.37, 16.22]
```

#### Windows 10 Copy Windows Defender Disabled

```
[7273.9684, 7427.1726, 7409.7312, 7417.4478, 7164.8717, 7427.4655, 7321.0481, 7292.2561, 7159.4540, 7120.8913]
```

**Comment by bestander** [comment]
That is some solid research, @keawade, thanks for sharing all the data.
The data suggests that the raw filesystem performance is the bottleneck for yarn installations on Windows.
I am not sure if Yarn can do anything here unless there is some smart copy command that works around the limitation

**Comment by kumarharsh** [comment]
@keawade thanks for taking so much pain compiling those numbers! @bestander could it be that since mom doesn't face these same problems while copying (perpetual scanning), maybe yarn is not signed? Could be that windows defender is not trusting yarn the same level as npm. Just a thought...

**Comment by bestander** [comment]
@kumarharsh, we'll need to measure the difference between npm and yarn then.
Maybe npm is copying less files (Yarn's hoisting is not optimized for smallest node_modules tree).
And it would be great if we could automatically whitelist yarn via installer.

**Comment by Daniel15** [comment]
> maybe yarn is not signed? Could be that windows defender is not trusting yarn the same level as npm.

I don't think scripts can be signed (with the exception of PowerShell scripts which do support Authenticode signatures), so I don't think Yarn and npm would differ in that regard. Yarn's installer is Authenticode signed just like npm's.

**Comment by Daniel15** [comment]
> And it would be great if we could automatically whitelist yarn via installer.

I feel like automatically touching a virus scanning whitelist would result in virus scanners marking the installer as malware. It seems like a risky thing to do. Perhaps we could automatically blacklist the directory in the search indexer, though.

**Comment by pongo** [comment]
@keawade I have tested robocopy with the options `/E /MT` (multi-threaded copies).

| Copy method          | Avg time |
|----------------------|----------|
| Copy-Item `-Recurse` | 20219    |
| Robocopy `/E`        | 26652    |
| Robocopy `/E /MT`    | 9043     |

**Raw data** (windows 10)

Copy-Item `-Recurse`

```
[19494.3827, 19471.0148, 19573.9441, 19896.9619, 19413.0355, 20050.4264, 19370.4315, 22959.5867, 20969.9693, 20994.3076]
```

Robocopy `/E`

```
[26522.4862, 26489.6131, 26654.8518, 26910.1073, 26536.042, 26836.0344, 26682.3544, 26408.4497, 26883.7998, 26605.5189]
```

Robocopy `/E /MT`

```
[9274.1374, 9125.6525, 9292.1629, 9014.8979, 8947.7882, 8985.4369, 8742.3616, 8915.4609, 8938.8326, 9200.9616]
```

**Comment by bestander** [comment]
> I don't think scripts can be signed (with the exception of PowerShell scripts which do support Authenticode signatures), so I don't think Yarn and npm would differ in that regard. Yarn's installer is Authenticode signed just like npm's.

Sounds reasonable.
Can we double check this?
If running `npm install` do Indexer and Defender show up in Task Manager?

**Comment by keawade** [comment]
## NPM Defender and Indexing Tests

I recorded the time it took to run `npm install` on [this repo](https://github.com/keawade/keawade.github.io/tree/vue) under the same set of conditions as the [Indexing and Defender Tests](https://github.com/yarnpkg/yarn/issues/990#issuecomment-277900118) for Yarn and Windows copy methods.

| OS                                          | Avg Time | Normalized |
|---------------------------------------------|----------|------------|
| Linux Mint (Yarn)                           |   5.5980 |  1.00000   |
| Linux Mint (NPM)                            |  28.9793 |  5.17672   |
| Windows 10 (No defender)                    |  42.6296 |  7.61514   |
| Windows 10 (No indexing)                    |  53.8791 |  9.62470   |
| Windows 10 (No defender, no indexing)       |  37.9727 |  6.78326   | 
| Windows 10 (No alterations)                 |  58.5047 | 10.45100   |

### Summary

Looks like NPM is also impacted by Windows Defender.

### Raw Data

#### NPM (Linux Mint)

```
[29.2353468, 35.6938315, 31.2105951, 30.9298704, 36.5016868, 31.8017671, 30.6387978, 32.3466556, 31.4340427]
```

#### NPM (Windows)

```
[61.2370640, 63.8799427, 62.3602369, 54.0541606, 55.1055082, 59.8259424, 56.7668692, 61.1153600, 54.7739699, 55.9277175]
```

#### NPM with Windows Defender Disabled

```
[41.1666621, 45.6951565, 43.1979249, 43.9185817, 40.8516877, 42.3445648, 43.5419790, 43.5084263, 45.0731120, 36.9975436]
```

#### NPM with Windows Indexing Disabled

```
[61.1470203, 58.6288137, 52.2553500, 52.4279906, 53.5446943, 54.2839412, 51.1620714, 52.1045756, 51.6424888, 51.5937462]
```

#### NPM with Windows Defender and Windows Indexing Disabled

```
[37.1311942, 37.7022530, 38.4630113, 37.5750357, 38.1434941, 37.2711589, 37.2249454, 39.4748951, 38.5522905, 38.1883537]
```

**Comment by bestander** [comment]
I suppose that we will need to work around the limitation of Windows being slower at disk IO by reducing the number of IO operations in general and educating people about Indexing and Defender.

**Comment by Daniel15** [comment]
> reducing the number of IO operations in general 

This is a good idea in general, and will help perf everywhere. It could also be pretty beneficial for people building on servers with slow hard drives.

**Comment by bestander** [comment]
Looking forward a PR to replace fs streams piping with a robocopy here https://github.com/yarnpkg/yarn/blob/master/src/util/fs.js#L301.

**Comment by kumarharsh** [comment]
Does anybody know why this happens for me (everytime)?

![image](https://cloud.githubusercontent.com/assets/760112/22819877/4384b2f4-ef9a-11e6-94c8-b3e5ef142249.png)

**Comment by bestander** [comment]
I think you have badly hoisted node_modules, this bug is going to be fixed in #2676

**Comment by destroyerofbuilds** [comment]
With @bestander's introduction of hardlinking in #2620 (Which works fine in Windows 7 without administrator privileges), my overall installation times, and `node_modules/` size, dropped.

Without hardlinking:

```
Done in 167.76s.

real    2m49.633s
user    0m0.229s
sys     0m1.368s

du -sh node_modules/
216M    node_modules/
```

With hardlinking:

```
Done in 58.07s.

real    0m59.967s
user    0m0.183s
sys     0m1.369s

du -sh node_modules/
189M    node_modules/
```

**Comment by bestander** [comment]
Wait for 0.21.1, it will have @kittens' fix to hoisting.
Should be even faster

**Comment by kuncevic** [comment]
I am on Win7 /w  Yarn v0.21.3
```
[3/4] Linking dependencies...
Done in 947.71s. 
```
Had to wait this amount of time adding any new package with `yarn add ...`

**Comment by bestander** [comment]
@kuncevic, what would be a clean install time for the project?
What is the size of node_modules folder in files?
How does it compare to npm install?

**Comment by kumarharsh** [comment]
@bestander this is the same problem with me. Any `yarn add` or `yarn remove` takes equal time - about , even after @kittens' hoisting fixes.

Everytime this happens:

1. First, the `Fetching packages` runs (takes about 30s):
![image](https://cloud.githubusercontent.com/assets/760112/24354918/730368d2-1311-11e7-8d68-d788154e1884.png)
2. Then, linking dependencies pauses  for about 1 or 2 minutes.
![image](https://cloud.githubusercontent.com/assets/760112/24354957/a0165bfe-1311-11e7-956d-7656621760da.png)
3. Then, it resumes with next step, and installs(?) 63k files again.
![image](https://cloud.githubusercontent.com/assets/760112/24354998/d46b3636-1311-11e7-82e6-c495558b9466.png)

As I said, this happens every single time I run `yarn add` or `yarn remove`. It does not matter if the dependency I'm installing depends on any other dependency, A simple `npm install` for installing a new dependency or upgrading an existing one takes a fraction of this time. Things did improve by 2x with @kittens' hoisting fixes, but still the time taken is too much.

@bestander if you want a reproducible case, please clone this repo: https://github.com/kumarharsh/yarn-bug, and run yarn install, and then `yarn add react-helmet`.

**Comment by bestander** [comment]
Yarn is preserving determinism every time it runs add/remove, so it needs to check if any dependency got hoisted to the root of node_modules when dependencies change.
That is why it runs full linking phase.

Fetching dependencies - download, you can't optimise it.
First linking phase (1561 operations) - it creates all the folders for all the dependencies.
Second linking phase (63K operations) - it copies the files from cache to node_modues.

Yarn optimises file copy operations by checking if the files are the same before doing the copy.
We might want to profile this area better and see if we can decrease number of unnecessary IO.
Maybe on Windows copying would be faster then checking?

What about npm, how fast it does clean install?

**Comment by kumarharsh** [comment]
A clean install for npm (`npm install`) takes `552301.1944ms`.
Installing an additional dependency (`npm install weird`) takes `57023.7593ms`. (Most of this time is wasted in paperjs trying to install canvas as a dep - but this time would be common for both npm & yarn)

A clean install for yarn (`yarn install`) takes `612698.4915ms`.
Installing an additional dependency (`yarn add weird`) takes `495633.0307ms`.

```
npm version 3.10.9
yarn version 0.21.3
```

**Comment by vbfox** [comment]
@bestander @kumarharsh Yarn doesn't optimises file copy operations on windows due to a nodejs bug (See #2958 for a potential fix in yarn code) that isn't present on node 7.1+ so you can get your second `yarn add` to be a lot faster just by upgrading node.

**Comment by kuncevic** [comment]
```
nvm install 7.8.0
npm install npm -g
nvm use 7.8.0
`git clone https://github.com/angular/material2`
cd material2
yarn install Done in 210.22s.
rimraf node_modules
yarn install Done in 180.66s.
rimraf node_modules
yarn install Done in 181.11s.
```

**Comment by Jkaveri** [comment]
A deadly project.

**Comment by vbfox** [comment]
@kuncevic Nice to see that upgrading node works for `add` :)

Regarding empty `node_modules` a good thing to do is to measure how much is due to yarn and how much is due to FileSystem, Hard drive & Anti virus.
What I did to test that was to copy the full node module of `material2` somewhere in yarn cache :

```batch
for /f "delims=" %i in ('yarn cache dir') do set yarncachedir=%i
xcopy /E /Y /I /Q node_modules %yarncachedir%\x-temp
```

And then fror each test I ran :
```batch
rd node_modules /s /q
powershell -Command "Measure-Command { xcopy /E /Y /I /Q %yarncachedir%\x-temp node_modules}"
```

And took the total seconds.

## Results

Here are the results on 2 PCs: My home PC & my work PC, home has ESET Nod32 that I can control, work has TrendMicro OfficeScan that I can't disable

||yarn install Home|xcopy Home|yarn install Work|xcopy Work
|-|-|-|-|-
|**AV Disabled**|34s|23s|-
|**AV Exclude cache & code**|38s|29s|-
|**Av Exclude cache only**|43s|31s|-
|**Av full**|48s|32s|100s|236s

Each time AV was enabled it was toping the CPU chart during `yarn install` or `xcopy` (On my home PC 30% cpu total was taken at the max but on my work PC it fill one core for xcopy & all my cores for yarn)
xcopy is slower on my work PC than yarn, I suspect because it doesn't copy files in parallel while yarn does (That shouldn't matter for IO bound operations but AV are making it a CPU bound operation & xcopy wasn't written to fight so much stupidity 😄 )

## In conclusion

Some changes could be done in yarn to sidestep that (symlinking files ?) but essentially coping lot of small files on windows, especially with an AV enabled is slow.

**Comment by szachara** [comment]
Adding npm, yarn cache folders and node.exe to defender's exclusion list would be enough, of course all this can't be in indexed folder. Now yarn add / rm takes 7 secs

**Comment by bestander** [comment]
Thanks everyone, a significant optimization for Windows landed in 0.24  https://github.com/yarnpkg/yarn/pull/3234#issuecomment-297552326

**Comment by cowwoc** [comment]
@vbfox Can you please add version numbers for `npm` and `yarn` in your benchmark?

**Comment by nebulr** [comment]
this is still a piece of shit for MacOSX

**Comment by thecjharries** [comment]
I'm still experiencing some crazy install times.  `yarn add` seems to install and link everything (all items in `package.json`, ~30k dependencies) all the time. 

Linux versions:
```sh-session
$ yarn -v
1.3.2
$ node -v
v8.9.3
```
Windows versions:
```sh-session
> conemu-cyg-64.exe --version
ConEmu cygwin/msys connector version 1.2.2
> wslbridge.exe --version
wslbridge 0.2.3
> Get-ItemProperty 'HKLM:\SOFTWARE\Microsoft\Windows NT\CurrentVersion' | Select-Object ProductName, CurrentMajorVersionNumber, CurrentMinorVersionNumber, ReleaseId, CurrentBuild, CurrentBuildNumber, BuildLabEx


ProductName               : Windows 10 Pro Insider Preview
CurrentMajorVersionNumber : 10
CurrentMinorVersionNumber : 0
ReleaseId                 : 1709
CurrentBuild              : 17025
CurrentBuildNumber        : 17025
BuildLabEx                : 17025.1000.amd64fre.rs_prerelease.171020-1626
```

I've got two (and a half) questions:

1. What's the accepted solution to the issue in this thread? Was it #3234 or tweaking Windows Defender?

   * If the solution was tweaking Windows Defender, is there a complete writeup of what to do somewhere?

2. Is my issue actually related to this thread, or should I create a new one?

**Comment by bestander** [comment]
Thanks for raising a new issue, I'll respond there

**Comment by mkamranhamid** [comment]
It's been almost an hour now and I'm waiting for this command to finish the process. I've followed the above points mentioned by @Altiano but nothing works

**Comment by mkamranhamid** [comment]
do we have any alternative for this? like can i use `npm i -g .` will it act in the same way or I'll have to make some changes because this code uses **yarn** `workspace`

**Comment by mkamranhamid** [comment]
So finally after struggling for 2-3 hours, I had to use `npm i -g .` instead of `yarn global add file:.`. npm worked like a charm

**Comment by robbykrlos** [comment]
Not sure how this is helping, but it is. Initially one change triggered a compilation that took ~1 minute. After running the steps below, it took 3-10 seconds. Not sure if it's still related to Windows Defender, of how fast the file access is after the complete delete and re-install (maybe it's like a defragmentation that helps on reading speeds?)

Install RimRaf:
``npm install rimraf -g``
And in the project folder delete the node_modules folder with:
``rimraf node_modules``
Install again all node_modules:
```
yarn install
yarn build
yarn serve
```
Build and Serve are now working normally 👍
